src/main/java/edu/cornell/kfs/module/ar/document/web/struts/CuContractsGrantsInvoiceDocumentAction.java

```diff
@@ -41,6 +57,138 @@ public ActionForward prorateBill(ActionMapping mapping, ActionForm form, HttpSer
         return super.prorateBill(mapping, form, request, response);
     }
 
+    @Override
+    public ActionForward route(ActionMapping mapping, ActionForm form, HttpServletRequest request, HttpServletResponse response) throws Exception {
+        ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument = ((ContractsGrantsInvoiceDocumentForm) form).getContractsGrantsInvoiceDocument();
+
+        String warningMessage = getContractsGrantsInvoiceDocumentWarningMessage(contractsGrantsInvoiceDocument);
+        if (StringUtils.isNotEmpty(warningMessage)) {
+            ActionForward forward = promptForFinalBillConfirmation(mapping, form, request, response, KFSConstants.ROUTE_METHOD, warningMessage, contractsGrantsInvoiceDocument);
+            if (forward != null) {
+                return forward;
+            }
+        }
+
+        return super.route(mapping, form, request, response);
+    }
+
+    protected String getContractsGrantsInvoiceDocumentWarningMessage(ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument) {
+        List<String> warningMessages = new ArrayList<>();
+        ConfigurationService configurationService = SpringContext.getBean(ConfigurationService.class);
+        if (contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().isFinalBillIndicator()) {
+            warningMessages.add(configurationService.getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_FINAL_BILL_INDICATOR));
+        }
+
+        String billingPeriod = contractsGrantsInvoiceDocument.getInvoiceGeneralDetail().getBillingPeriod();
+        if (StringUtils.length(billingPeriod) != CuArConstants.CINV_DATE_RANGE_EXPECTED_FORMAT_LENGTH) {
+            billingPeriod = StringUtils.defaultString(billingPeriod);
+            String warningMessage = SpringContext.getBean(ConfigurationService.class).getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_DATE_RANGE_INVALID_FORMAT_LENGTH);
+            warningMessages.add(MessageFormat.format(warningMessage, new String[] {billingPeriod}));
+        } else {
+            try {
+                validateBillingPeriodDateRange(contractsGrantsInvoiceDocument, warningMessages);
+            } catch(ParseException ex) {
+                LOG.error("getContractsGrantsInvoiceDocumentWarningMessage: " + ex.getMessage());
+                String warningMessage = SpringContext.getBean(ConfigurationService.class).getPropertyValueAsString(CUKFSKeyConstants.WARNING_CINV_DATE_RANGE_PARSE_EXCEPTION);
+                warningMessages.add(warningMessage + CuArConstants.QUESTION_NEWLINE_STRING + ex.toString());
+            }
+        }
+
+        return StringUtils.join(warningMessages, CuArConstants.QUESTION_NEWLINE_STRING);
+    }
+
+    protected void validateBillingPeriodDateRange(ContractsGrantsInvoiceDocument contractsGrantsInvoiceDocument, List<String> warningMessages) throws ParseException {
```

Comment:
To keep code duplication to a minimum while still keeping the method length reasonable, I would suggest taking the code from the three date validation methods below and combining them into this "validateBillingPeriodDateRange" method. That way, you don't have to keep repeating the variable initialization for start/end dates or the service references.